examples/advanced/lr-newton-raphson/job/newton_raphson/app/custom/newton_raphson_train.py

```diff
@@ -0,0 +1,229 @@
+from nvflare.app_common.abstract.fl_model import FLModel, ParamsType
+
+
+def parse_arguments():
+  """
+  Parse command line args for client side training.
+  """
+  parser = argparse.ArgumentParser(
+    description="Federated Second-Order Newton Raphson"
+  )
+
+  parser.add_argument(
+    '--data_root',
+    type=str,
+    help='Path to load client side data.'
+  )
+
+  return parser.parse_args()
+
+
+def load_data(
+    data_root,
+    site_name
+):
+  """
+  Load the data for each client.
+
+  Args:
+      data_root: root directory storing client site data.
+      site_name: client site name
+  Returns:
+      A dict with client site training and validation data.
+  """
+  print('loading data for client {} from: {}'.format(site_name, data_root))
+  train_x_path = os.path.join(data_root, '{}.train.x.npy'.format(site_name))
+  train_y_path = os.path.join(data_root, '{}.train.y.npy'.format(site_name))
+  test_x_path = os.path.join(data_root, '{}.test.x.npy'.format(site_name))
+  test_y_path = os.path.join(data_root, '{}.test.y.npy'.format(site_name))
+
+  train_X = np.load(train_x_path)
+  train_y = np.load(train_y_path)
+  valid_X = np.load(test_x_path)
+  valid_y = np.load(test_y_path)
+
+  return {
+    'train_X': train_X,
+    'train_y': train_y,
+    'valid_X': valid_X,
+    'valid_y': valid_y
+  }
+
+
+def sigmoid(inp): return 1.0 / (1.0 + np.exp(-inp))
+
+
+def train_newton_raphson(data, theta):
+  """
+  Compute gradient and hessian on local data
+  based on paramters received from server.
+
+  """
+  train_X = data['train_X']
+  train_y = data['train_y']
+
+  # Add intercept, pre-pend 1s to as first
+  # column of train_X
+  train_X = np.concatenate(
+    (np.ones((train_X.shape[0], 1)), train_X),
+    axis=1
+  )
+
+  # Compute probabilities from current weights
+  proba = sigmoid(np.dot(train_X, theta))
+
+  # The gradient is X^T . (y - proba)
+  gradient = np.dot(train_X.T, (train_y - proba))
+
+  # The hessian is X^T . D . X, where D is the
+  # diagnoal matrix with values proba * (1 - proba)
+  D = np.diag((proba * (1 - proba))[:,0])
+  hessian = train_X.T.dot(D).dot(train_X)
+
+  return {
+    'gradient': gradient,
+    'hessian': hessian
+  }
+
+
+def validate(data, theta):
+  """
+  Performs local validation.
+  Computes accuracy and precision scores.
+
+  """
+  valid_X = data['valid_X']
+  valid_y = data['valid_y']
+
+  # Add intercept, pre-pend 1s to as first
+  # column of valid_X
+  valid_X = np.concatenate(
+    (np.ones((valid_X.shape[0], 1)), valid_X),
+    axis=1
+  )
+
+  # Compute probabilities from current weights
+  proba = sigmoid(np.dot(valid_X, theta))
+
+  return {
+    'accuracy': accuracy_score(valid_y, proba.round()),
+    'precision': precision_score(valid_y, proba.round())
+  }
+
+
+def main():
+  """
+  This is a typical ML training loop,
+  augmented with Flare Client API to
+  perform local training on each client
+  side and send result to server.
```

Comment:
some explanations (in readme) by comparing with train_centralized?

Reply:
Add 1 paragraph at the end of "Implementation" to explain a bit the different between federated training code vs centralized one. Could you take a look?

Reply:
@ZiyueXu77